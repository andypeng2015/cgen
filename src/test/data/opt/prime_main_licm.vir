module prime_main_licm

const data $fmt = {
  "prime(%d) = %d\n",
  0x0_b
}

export function w $main(w %argc, l %argv) {
@start:
  %p = add.l %argv, 0x8_l
  %p = ld.l %p
  %n = call.w $atoi(%p)
  %m = copy.w 1_w
  %i = copy.w 1_w
  jmp @loop1
@loop1:
  %c = sle.w %n, 0_w
  br %c, @done, @loop2
@loop2:
  %b = call.w $is_prime(%i)
  %c = eq.w %b, 0_w
  br %c, @no, @yes
@yes:
  %k = call.w $printf($fmt, ..., %m, %i)
  %n = sub.w %n, 1_w
  %i = add.w %i, 1_w
  %m = add.w %m, 1_w
  jmp @loop1
@no:
  %i = add.w %i, 1_w
  jmp @loop2
@done:
  ret 0_w
}
