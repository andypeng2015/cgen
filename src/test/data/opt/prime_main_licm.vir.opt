module prime_main_licm

const data $fmt = {
  "prime(%d) = %d\n",
  0x0_b
}

export function w $main(w %argc, l %argv) {
@2:
  %6 = add.l %argv, 0x8_l ; @27
  %p.2 = ld.l %6 ; @5
  %n.1 = call.w $atoi(%p.2) ; @6
  jmp @3(%n.1, 0x1_w, 0x1_w)
@3(%n.2, %m.2, %i.2):
  %7 = sle.w %n.2, 0x0_w ; @28
  br %7, @9, @10(%i.2)
@10(%i.3):
  %b.1 = call.w $is_prime(%i.3) ; @14
  %8 = add.w %i.3, 0x1_w ; @29
  %9 = eq.w %b.1, 0x0_w ; @30
  br %9, @10(%8), @13
@13:
  %k.1 = call.w $printf($fmt, ..., %m.2, %i.3) ; @16
  %10 = sub.w %n.2, 0x1_w ; @31
  %11 = add.w %m.2, 0x1_w ; @32
  jmp @3(%10, %11, %8)
@9:
  ret 0x0_w
}
