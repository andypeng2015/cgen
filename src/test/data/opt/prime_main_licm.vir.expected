module prime_main_licm

const data $fmt = {
  "prime(%d) = %d\n",
  0x0_b
}

export function w $main(w %argc, l %argv) {
@2:
  %0 = add.l %argv, 0x8_l ; @21
  %p.2 = ld.l %0 ; @5
  %n.1 = call.w $atoi(%p.2) ; @6
  jmp @3(%n.1, 0x1_w, 0x1_w)
@3(%n.2, %m.2, %i.2):
  %1 = sle.w %n.2, 0x0_w ; @22
  br %1, @9, @10(%i.2)
@10(%i.3):
  %b.1 = call.w $is_prime(%i.3) ; @14
  %3 = eq.w %b.1, 0x0_w ; @24
  %2 = add.w %i.3, 0x1_w ; @23
  br %3, @10(%2), @13
@13:
  %k.1 = call.w $printf($fmt, ..., %m.2, %i.3) ; @16
  %4 = sub.w %n.2, 0x1_w ; @25
  %5 = add.w %m.2, 0x1_w ; @26
  jmp @3(%4, %5, %2)
@9:
  ret 0x0_w
}
